### PR TITLE
[mds-attachment-service] Add attachment_list_id column to attachments table

### PR DESCRIPTION
**Pull request by avatarneil**
## 📚 Purpose
Adds an `attachment_list_id` column to the attachments table, so it can be used for grouping attachments together, as well as linking them to a foreign object (e.g. an `audit` via its `audit_trip_id`).

Also snuck in a slight tweak to the `SerializedBuffers` type to try to make it a tad more readable.

## 📦 Impacts:
- mds-attachment-service
- mds-types

### CODE REVIEW

File: packages/mds-attachment-service/@types/index.ts [hunk 1]
```diff
@@ -3,23 +3,40 @@ import { RpcServiceDefinition, RpcRoute } from '@mds-core/mds-rpc-common'
...
 
 export interface AttachmentService {
   deleteAttachment: (attachment_id: UUID) => AttachmentDomainModel
-  writeAttachment: (file: SerializedBuffers<Express.Multer.File>) => AttachmentDomainModel
+  writeAttachment: (
+    file: SerializedBuffers<Express.Multer.File>,
+    attachment_list_is: Nullable<UUID>
+  ) => AttachmentDomainModel
+  readAttachment: (attachment_id: UUID) => AttachmentDomainModel | undefined
+  readAttachments: (options: ReadAttachmentsOptions) => AttachmentDomainModel[]
```
Comment (drtyh2o) [inline packages/mds-attachment-service/@types/index.ts:34]: Just for comparison, what if the service methods were:
```ts
readAttachment: (attachment_id: UUID) => AttachmentDomainModel | undefined
readAttachments: (attachment_ids: UUID[]) => AttachmentDomainModel[]
readAttachmentsList: (attachment_list_id: UUID) => AttachmentDomainModel[]
```

Comment (avatarneil) [reply packages/mds-attachment-service/@types/index.ts:34]: Yeah we could definitely do that, I'd lean towards calling the `readAttachments` method` readAttachmentsByIds` or something

Comment (drtyh2o) [reply packages/mds-attachment-service/@types/index.ts:34]: How do you reconcile this:
```ts
readAttachment('uuid_that_does_not_exist') => undefined
```
vs
```ts
readAttachments(['uuid_that_does_not_exist']) => [] // presumably
```

Is that OK or should readAttachments return `(AttachmentDomainModel | undefined)[]`?

File: packages/mds-attachment-service/repository/index.ts [hunk 2]
```diff
@@ -51,6 +51,54 @@ class AttachmentReadWriteRepository extends ReadWriteRepository {
...
+      throw RepositoryError(error)
+    }
+  }
+
+  public readAttachments = async (options: ReadAttachmentsOptions): Promise<AttachmentDomainModel[]> => {
+    const isAttachmentListIdOption = (opts: ReadAttachmentsOptions): opts is { attachment_list_id: UUID } => {
+      if ('attachment_list_id' in opts) return true
+
+      return false
+    }
```
Comment (drtyh2o) [inline packages/mds-attachment-service/repository/index.ts:72]:
```suggestion
    const isAttachmentListIdOption = (opts: ReadAttachmentsOptions): opts is { attachment_list_id: UUID } => 'attachment_list_id' in opts
```

File: packages/mds-attachment-service/repository/index.ts [hunk 3]
```diff
@@ -51,6 +51,54 @@ class AttachmentReadWriteRepository extends ReadWriteRepository {
...
+
+    const isAttachmentIdsOption = (opts: ReadAttachmentsOptions): opts is { attachment_ids: UUID[] } => {
+      if ('attachment_ids' in opts) return true
+
+      return false
+    }
```
Comment (drtyh2o) [inline packages/mds-attachment-service/repository/index.ts:78]:
```suggestion
    const isAttachmentIdsOption = (opts: ReadAttachmentsOptions): opts is { attachment_ids: UUID[] } => 'attachment_ids' in opts
```

File: packages/mds-attachment-service/@types/index.ts (hunk 1 above)
Comment (avatarneil) [reply packages/mds-attachment-service/@types/index.ts:34]: It _probably_ ought to return a `(AttachmentDomainModel | undefined)[]`, but the proper solution to that design goal I think is the new `--noUncheckedIndexedAccess` operator. Any array of things could possibly be empty, unless it's a tuple

**Review by drtyh2o — APPROVED** [review]

File: packages/mds-attachment-service/@types/index.ts [hunk 4]
```diff
@@ -3,23 +3,40 @@ import { RpcServiceDefinition, RpcRoute } from '@mds-core/mds-rpc-common'
...
 }
 
 export type AttachmentDomainCreateModel = DomainModelCreate<AttachmentDomainModel>
 
 export interface AttachmentService {
   deleteAttachment: (attachment_id: UUID) => AttachmentDomainModel
-  writeAttachment: (file: SerializedBuffers<Express.Multer.File>) => AttachmentDomainModel
+  writeAttachment: (
+    file: SerializedBuffers<Express.Multer.File>,
+    attachment_list_is: Nullable<UUID>
```
Comment (BigIdeas) [inline packages/mds-attachment-service/@types/index.ts:31]: typo: `attachment_list_is` → `attachment_list_id`

Comment (avatarneil) [reply packages/mds-attachment-service/@types/index.ts]: Great catch, thanks!

**Review by BigIdeas — APPROVED** [review]

**Pull request merged**